Orange/widgets/visualize/tests/test_owmosaic.py

```diff
@@ -235,58 +235,43 @@ def test_color_combo(self):
             discrete_data = self.widget.discrete_data
 
             if color == "(Pearson residuals)":
+                self.widget.interior_coloring = self.widget.PEARSON
                 self.assertIsNone(discrete_data.domain.class_var)
             else:
+                self.widget.interior_coloring = self.widget.CLASS_DISTRIBUTION
                 self.assertEqual(color, str(discrete_data.domain.class_var))
+
             output = self.get_output("Data")
             self.assertEqual(output.domain.class_var, table.domain.class_var)
 
-    def test_vizrank(self):
+            for ma in range(2,5):
+                self.vizrank.max_attrs = ma
+                sc = self.vizrank.state_count()
+                self.assertTrue((sc == 6 and ma == 2) or
+                                (ma > 2 and sc == (7 + ((i > 0) - 1) * (-ma))))
+
+    def test_scores(self):
         """
-        Tests if vizrank works accordingly to selected color value.
-        GH-2133
+        Test scores without running vizrank.
+        GH-2299
         GH-2036
         """
-        def test(i):
-            self.send_signal("Data", table)
-            idx = self.widget.cb_attr_color.findText(color_vars[i])
-            self.widget.cb_attr_color.setCurrentIndex(idx)
-            color = self.widget.cb_attr_color.currentText()
-            simulate.combobox_activate_index(self.widget.controls.variable_color, idx, 0)
-            self.widget.vizrank.button.click()
-            time.sleep(.5)
-            model = self.widget.vizrank.rank_model
-            nrows = model.rowCount()
-            all_data = []
-            self.assertTrue((nrows == 7 and i != 0) or (nrows == 10 and i == 0))
-            for row in range(nrows):
-                data = []
-                for column in range(model.columnCount()):
-                    index = model.index(row, column)
-                    data.append(str(model.data(index)))
-                    all_data = all_data + str(model.data(index)).replace(',', '').split()
-                for text in data:
-                    self.assertNotIn(color, text)
-            self.assertEqual(len(set(all_data)), 3 if color != "(Pearson residuals)" else 4)
-            self.vizrank.stop_and_reset()
-            self.vizrank.toggle()
-
+        SCORES = [4.35e-40, 2.82e-05, 6.18e-100, 1.06e-123, 4.15e-47, 5.49e-102, 5.30e-128]
         table = Table("titanic")
         self.send_signal("Data", table)
-        self.widget.vizrank.max_attrs = 4
-        color_vars = ["(Pearson residuals)"] + [str(x) for x in table.domain]
-
-        for i in range(len(color_vars)):
-            self.widget = self.create_widget(OWMosaicDisplay)
-            test(i)
+        self.vizrank.compute_attr_order()
+        self.widget.vizrank.max_attrs = 3
+        state = None
+        for state, score in zip(self.vizrank.iterate_states(state), SCORES):
+            self.assertAlmostEqual(self.vizrank.compute_score(state), score)
```

Comment:
I think that by default, `assertAlmostEqual` compares to 7 decimal places. This could be problematic with scores so close to 0...